packages/cozy-client/src/store/documents.js

```diff
@@ -188,38 +185,65 @@ export const getCollectionFromSlice = (state = {}, doctype) => {
   return Object.values(state[doctype])
 }
 
-/*
-  This method has been created in order to get a returned object
-  in `data` with the full set on information coming potentially from
+/**
+ * 
+ * This method has been created in order to get a returned object
+   in `data` with the full set on information coming potentially from
   `included`
-
+ 
   This method should be somewhere else. The `document` shall not be
   dealt with included / data and so on.
-
+ 
   This method takes `data` and `included` and merge both sources
   together. It should be always up-to-date. The returned object
   will be as full of information as it can be.
-*/
-export const extractAndMergeDocument = (data, updatedStateWithIncluded) => {
-  const doctype = data[0]._type
+ 
+ * @param {Array<import('../types').CozyClientDocument>} newData - Queries documents 
+ * @param {import("../types").DocumentsStateSlice} updatedStateWithIncluded documents' state 
+ * @returns {import("../types").DocumentsStateSlice} the merged documents state
+ */
+export const extractAndMergeDocument = (newData, updatedStateWithIncluded) => {
+  const doctype = newData[0]._type
 
   if (!doctype) {
-    logger.info('Document without _type', data[0])
+    logger.info('Document without _type', newData[0])
     throw new Error('Document without _type')
   }
-  const sortedData = keyBy(data, properId)
-
-  let mergedData = Object.assign({}, updatedStateWithIncluded)
-  mergedData[doctype] = Object.assign({}, updatedStateWithIncluded[doctype])
 
-  Object.values(sortedData).map(data => {
-    const id = properId(data)
-    if (mergedData[doctype][id]) {
-      mergedData[doctype][id] = merge({}, mergedData[doctype][id], data)
-    } else {
-      mergedData[doctype][id] = data
+  // Index docs by id, for faster retrieval
+  const newDocsMap = new Map(newData.map(doc => [properId(doc), doc]))
+
+  const mergedState = Object.assign({}, updatedStateWithIncluded)
+  // FIXME: this was introduced for a bug in banks, but it does not look useful
+  mergedState[doctype] = Object.assign({}, updatedStateWithIncluded[doctype])
+  if (!mergedState[doctype]) {
+    mergedState[doctype] = {}
+  }
+
+  let haveDocumentsChanged = false
+
+  newDocsMap.forEach((newDoc, key) => {
+    const id = key
+    const currentDoc = mergedState[doctype][id]
+    if (!currentDoc) {
+      mergedState[doctype][id] = newDoc
+      haveDocumentsChanged = true
+    } else if (!fastEqual(currentDoc, newDoc)) {
+      // XXX Here we check if the doc has changed. Note this would be way faster
```

Comment:
nit: What are theses XXX ?

Reply:
This is something I took from @nono , I'm not sure if it means anything, but I interpret it as "this is tricky and worth reading"

Reply:
Yes, it means that it requires attention, and often, it is dirty (like we do something weird to avoid a bug in a lower level library or we have to do that for an historical reason).